### PR TITLE
Reusable `Payload`

### PR DESCRIPTION
**Pull request by NiteshKant**
__Problem__

`PayloadImpl` caches `data` and `metadata` buffers. This means once those buffers are read, they can not be reused as the pointer of `ByteBuffer` has moved to the end of buffer.
This restricts the usage of `PayloadImpl` instance with something like:

```java
socket.requestResponse(new PayloadImpl("Hello")).retry(2);
```

For any retry in the above code, data read from the payload instance will be empty.

__Modification__

- store buffer `position()` on creation and reset them on every `get()`.
- Additional constructor to override this behavior and create a payload for single use.
- `PayloadImpl` defaults to reusable payload.

__Result__

Possible to retry a request without having a custom `Payload` implementation.

### CODE REVIEW

File: reactivesocket-core/src/main/java/io/reactivesocket/util/PayloadImpl.java [hunk 1]
```diff
@@ -61,17 +71,39 @@ public PayloadImpl(ByteBuffer data) {
...
+     * New instance where every invocation of {@link #getMetadata()} and {@link #getData()} will reset the position of
+     * the buffer to the position when it is created, if and only if, {@code singleUse} is set to {@code false}.
+     *
+     * @param data Buffer for data.
+     * @param metadata Buffer for metadata.
+     * @param singleUse {@code true} if the buffer position is to be reset on every invocation of {@link #getData()} and
+     * {@link #getMetadata()}.
+     */
+    public PayloadImpl(ByteBuffer data, ByteBuffer metadata, boolean singleUse) {
         this.data = data;
```
Comment (NiteshKant) [inline reactivesocket-core/src/main/java/io/reactivesocket/util/PayloadImpl.java]: @robertroeser `ByteBuffer.slice()` always uses the current position i.e. there is no such method as `ByteBuffer.slice(position)`. So, if we were to slice, we still have to reset the position first.
Do you see any other value in using `slice()` even though we have to reset position first?

File: reactivesocket-core/src/main/java/io/reactivesocket/util/PayloadImpl.java [hunk 2]
```diff
@@ -61,17 +71,39 @@ public PayloadImpl(ByteBuffer data) {
...
+    /**
+     * New instance where every invocation of {@link #getMetadata()} and {@link #getData()} will reset the position of
+     * the buffer to the position when it is created, if and only if, {@code singleUse} is set to {@code false}.
+     *
+     * @param data Buffer for data.
+     * @param metadata Buffer for metadata.
+     * @param singleUse {@code true} if the buffer position is to be reset on every invocation of {@link #getData()} and
+     * {@link #getMetadata()}.
+     */
+    public PayloadImpl(ByteBuffer data, ByteBuffer metadata, boolean singleUse) {
```
Comment (stevegury) [inline reactivesocket-core/src/main/java/io/reactivesocket/util/PayloadImpl.java]: nitpick: but renaming/reversing `singleUse` to `reusable` may be easier to understand (especially if you don't read the doc).

File: reactivesocket-core/src/main/java/io/reactivesocket/util/PayloadImpl.java (hunk 1 above)
Comment (NiteshKant) [inline reactivesocket-core/src/main/java/io/reactivesocket/util/PayloadImpl.java]: ping @robertroeser

**Comment by ktoso** [comment]
Ping, any chance to get this in soon? :)

**Comment by NiteshKant** [comment]
@ktoso you got it 😄

File: reactivesocket-core/src/main/java/io/reactivesocket/util/PayloadImpl.java (hunk 2 above)
Comment (NiteshKant) [inline reactivesocket-core/src/main/java/io/reactivesocket/util/PayloadImpl.java]: @stevegury done, renamed to `reusable` and changed implementation!

**Comment by ktoso** [comment]
Excellent, thanks! Very nice to see much movement here in the new year 👍

**Pull request merged**